File: 백준/Gold/1027. 고층 건물/고층 건물.py
import sys

N = int(sys.stdin.readline())

buildings = list(map(int, sys.stdin.readline().split()))
count = 0
INF = int(1e9)

for i in range(N):
    x1, y1 = i, buildings[i]
    cnt = 0
    prev = -INF

    for j in range(i + 1, N):
        x2, y2 = j, buildings[j]

        slope = (y2 - y1) / (x2 - x1)

        if slope > prev:
            cnt += 1
            prev = slope

    prev = INF
    
    for k in range(i - 1, -1, -1):
        x3, y3 = k, buildings[k]

        slope = (y1 - y3) / (x1 - x3)

        if slope < prev:
            cnt += 1
            prev = slope

    count = max(count, cnt)

print(count)

File: 백준/Gold/1027. 고층 건물/고층 건물.js
const filePath = process.platform === 'linux' ? '/dev/stdin' : './Javascript/input.txt';
const input = require('fs').readFileSync(filePath).toString().trim().split('\n');
const N = +input[0];
const buildings = input[1].split(' ').map(Number);
const views = Array(N).fill(0);

for (let i = 0; i < N; i++) {
    let left = Infinity;
    let right = -Infinity;
    const x1 = i;
    const y1 = buildings[i];
    for (let j = i - 1; j >= 0; j--) {
        const x2 = j;
        const y2 = buildings[j];
        const incline = (y2 - y1) / (x2 - x1);
        if (incline < left) {
            left = incline;
            views[i]++;
        }
    }
    for (let j = i + 1; j < N; j++) {
        const x2 = j;
        const y2 = buildings[j];
        const incline = (y2 - y1) / (x2 - x1);
        if (incline > right) {
            right = incline;
            views[i]++;
        }
    }
}

console.log(Math.max(...views));
